apollo-router/src/services/connector_service/snapshot.rs

```diff
@@ -0,0 +1,122 @@
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use http::HeaderMap;
+use http::HeaderName;
+use http::HeaderValue;
+use indexmap::IndexMap;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_json_bytes::Value;
+use tracing::warn;
+
+use crate::services::router::body::RouterBody;
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct Snapshot<'a> {
+    url: Cow<'a, str>,
+    headers: IndexMap<String, String>,
+    body: Cow<'a, Value>,
+}
+
+impl<'a> Snapshot<'a> {
+    pub(crate) fn new(url: &'a str, body: &'a Value, headers: &'a HeaderMap<HeaderValue>) -> Self {
+        let headers = headers
+            .iter()
+            .map(|(k, v)| (k.to_string(), v.to_str().unwrap().to_string()))
+            .collect();
+        Snapshot {
+            url: Cow::Borrowed(url),
+            headers,
+            body: Cow::Borrowed(body),
+        }
+    }
+
+    pub(crate) fn save(&self, base: PathBuf) -> Result<(), Box<dyn std::error::Error>> {
+        let path = snapshot_path(base.clone(), &self.url);
+        if let Some(parent) = path.parent() {
+            std::fs::create_dir_all(parent)?;
+        }
+        std::fs::write(path, serde_json::to_string_pretty(&self)?)?;
+        Ok(())
+    }
+
+    pub(crate) fn load(base: PathBuf, url: &'a str) -> Option<Self> {
+        let path = snapshot_path(base, url);
+        let string = std::fs::read_to_string(path).ok()?;
+        let snapshot: Snapshot = serde_json::from_str(&string).ok()?;
+        if snapshot.url != url {
+            // TODO: handle URL collisions
+            warn!("Snapshot URL collision: {}, {}", snapshot.url, url);
+            return None;
+        }
+        Some(snapshot)
+    }
+}
+
+impl<'a> TryFrom<Snapshot<'a>> for crate::plugins::connectors::http::Result<RouterBody> {
+    type Error = ();
+
+    fn try_from(snapshot: Snapshot) -> Result<Self, Self::Error> {
+        let mut response = http::Response::builder().status(http::StatusCode::OK);
+        if let Some(headers) = response.headers_mut() {
+            for (name, value) in snapshot.headers.into_iter() {
+                if let Ok(name) = HeaderName::from_str(&name.clone()) {
+                    if let Ok(value) = HeaderValue::from_str(&value.clone()) {
+                        headers.insert(name, value);
+                    }
+                }
+            }
+        }
+        if let Ok(string) = serde_json::to_string(&*snapshot.body) {
+            if let Ok(response) = response.body(RouterBody::from(string)) {
+                return Ok(crate::plugins::connectors::http::Result::HttpResponse(
+                    response,
+                ));
+            }
+        }
+        Err(())
+    }
+}
+
+fn snapshot_path(base: PathBuf, url: &str) -> PathBuf {
+    let url_hash = {
+        let mut hasher = DefaultHasher::new();
+        url.hash(&mut hasher);
+        hasher.finish()
+    };
+    base.join(url_hash.to_string()).with_extension("json")
```

Comment:
we might need to hash the body too. i've seen apis like `POST /search` where all the unique parameters are in the body

Reply:
Interesting - that could be a good improvement as long as the body doesn't contain any data that might change, like a timestamp or something. I also realize I should probably include HTTP method in case somebody uses the same URL for PUT/POST or something.